src/Upcast/Infra/Types/Resource.hs

```diff
@@ -150,512 +111,111 @@ class Resource a where
-    asks ctxTags >>= toEc2Tags [subnet] . (("hash", hashOf Ec2subnet{..}):) . (("Name", ec2subnet_name):)
-    return subnet
-  update :: AWS m => ResourceId -> Ec2subnet -> m ResourceId
-  update current Ec2subnet{..} = do
-    -- If CIDRs overlap, we needs to delete *now* to avoid conflict on 'create'.
-    currentCIDR <- EC2.describeSubnets
-                 & EC2.dsFilters .~ [EC2.filter' "subnet-id" & EC2.fValues .~ [current]]
-                 & send <&> read . unpack . view (EC2.dsrsSubnets . folded . EC2.subCIdRBlock)
-    let canDefer = currentCIDR `overlaps` read (unpack ec2subnet_cidrBlock)
-    EC2.deleteSubnet current & (if canDefer then void . defer else void . return) . send
-    create Ec2subnet{..}
-  reify :: Map Reference ResourceId -> Ec2subnet -> Either Missing Ec2subnet
-  reify (lookup_ -> lookup_) Ec2subnet{..} = do
-    ec2subnet_vpc <- lookup_ (Reference "ec2-vpc") ec2subnet_vpc
-    return Ec2subnet{..}
-
-keyPrefix :: Tags -> Ec2keypair -> Text
-keyPrefix tags Ec2keypair{..} = T.intercalate "-" $
-  [ tags ! "created-using"
-  , tags ! "realm"
-  , tags ! "expression"
-  , ec2keypair_name
-  ] where (!) = fmap (fromMaybe "") . flip lookup
+  create = Subnet.create
+  update = Subnet.update
+  reify = Subnet.reify
 
 instance Matcher Ec2keypair EC2.KeyPairInfo where
   type Rq Ec2keypair = EC2.DescribeKeyPairs
-  request = \f s -> const s <$> f EC2.describeKeyPairs
-  candidates Ec2keypair{..} keypairs =  do
-    prefix <- flip keyPrefix Ec2keypair{..} <$> asks ctxTags
-    keypairs ^.. EC2.dkprsKeyPairs . folded . filtered (T.isPrefixOf prefix . view extractId) & return
-  hashesTo keypair = T.isSuffixOf (hashOf keypair) . view (EC2.kpiKeyName . _Just)
-  -- XXX: If 'Nothing', "" by mempty on Traversal, don't know how bad this is.
-  extractId = EC2.kpiKeyName . _Just
+  request = Keypair.request
+  candidates = Keypair.candidates
+  hashesTo = Keypair.hashesTo
+  extractId = Keypair.extractId
 
 instance AWSPager EC2.DescribeKeyPairs where page _ _ = Nothing
 
-addKeypair :: AWS m => Ec2keypair -> m ()
-addKeypair key = modify $ \State{..} -> State{..} { stateKeyPairs = key : stateKeyPairs }
-
 instance Resource Ec2keypair where
-  create :: AWS m => Ec2keypair -> m ResourceId
-  create Ec2keypair{..} = do
-    name <- (<> "-" <> hashOf Ec2keypair{..}) . flip keyPrefix Ec2keypair{..} <$> asks ctxTags
-    -- XXX: This filename interface leads to bad hashing; contents can change without updates registering.
-    pubkey <- liftIO . expectRight . fgconsume $ exec "ssh-keygen" ["-f", unpack ec2keypair_privateKeyFile, "-y"]
-    EC2.importKeyPair name pubkey & send <&> view (EC2.ikprsKeyName . _Just)
-  update :: AWS m => ResourceId -> Ec2keypair -> m ResourceId
-  update current Ec2keypair{..} = do
-    keypair <- create Ec2keypair{..}
-    EC2.deleteKeyPair current & defer . send
-    return keypair
+  create = Keypair.create
+  update = Keypair.update
+
   match Ec2keypair{..} = do
```

Comment:
It seems that bit of state has nothing to do with the actual "cloud ops", so i left the state code there.